worker/remoterelations/remoterelations.go

```diff
@@ -388,86 +369,117 @@ func (w *remoteApplicationWorker) relationChanged(
+		return errors.Trace(err)
+	}
+	relationUnitsWatcher, err := newRelationUnitsWatcher(
+		relationTag,
+		w.relationInfo.applicationId,
+		remoteRelationId,
+		localRelationUnitsWatcher,
+		w.relationChanges,
+		w.facade,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(relationUnitsWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	r := &relation{}
+	r.RelationId = remoteRelation.Id
+	r.Life = remoteRelation.Life
+	r.ruw = relationUnitsWatcher
+	relations[key] = r
 	return nil
 }
 
 func (w *remoteApplicationWorker) registerRemoteRelation(
-	relationTag names.Tag, publisher RemoteRelationChangePublisher,
-) (params.RemoteEntityId, error) {
+	applicationTag, relationTag names.Tag, publisher RemoteRelationChangePublisher,
+) (remoteApplicationId params.RemoteEntityId, remoteRelationId params.RemoteEntityId, _ error) {
 	logger.Debugf("register remote relation %v", relationTag.Id())
+	emptyId := params.RemoteEntityId{}
 	// Ensure the relation is exported first up.
-	results, err := w.facade.ExportEntities([]names.Tag{relationTag})
+	results, err := w.facade.ExportEntities([]names.Tag{applicationTag, relationTag})
 	if err != nil {
-		return params.RemoteEntityId{}, errors.Annotatef(err, "exporting relation %v", relationTag)
+		return emptyId, emptyId, errors.Annotatef(err, "exporting relation %v and application", relationTag, applicationTag)
 	}
 	if results[0].Error != nil && !params.IsCodeAlreadyExists(results[0].Error) {
-		return params.RemoteEntityId{}, errors.Annotatef(err, "exporting relation %v", relationTag)
+		return emptyId, emptyId, errors.Annotatef(err, "exporting application %v", applicationTag)
+	}
+	remoteApplicationId = *results[0].Result
+	if results[1].Error != nil && !params.IsCodeAlreadyExists(results[1].Error) {
+		return emptyId, emptyId, errors.Annotatef(err, "exporting relation %v", relationTag)
 	}
-	remoteRelationId := *results[0].Result
+	remoteRelationId = *results[1].Result
 
 	// This data goes to the remote model so we map local info
 	// from this model to the remote arg vales and visa versa.
 	arg := params.RegisterRemoteRelation{
-		ApplicationId:          w.relationInfo.applicationId,
+		ApplicationId:          remoteApplicationId,
 		RelationId:             remoteRelationId,
 		RemoteEndpoint:         w.relationInfo.localEndpoint,
 		OfferedApplicationName: w.relationInfo.remoteApplicationName,
 		LocalEndpointName:      w.relationInfo.remoteEndpointName,
 	}
-	if err := publisher.RegisterRemoteRelation(arg); err != nil {
-		return params.RemoteEntityId{}, errors.Trace(err)
+	remoteAppIds, err := publisher.RegisterRemoteRelations(arg)
+	if err != nil {
+		return emptyId, emptyId, errors.Trace(err)
+	}
+	if err := results[0].Error; err != nil && !params.IsCodeAlreadyExists(err) {
+		return emptyId, emptyId, errors.Annotatef(err, "registering relation %v", relationTag)
+	}
+	// Import the application id from the offering model.
+	offeringRemoteAppId := remoteAppIds[0].Result
```

Comment:
And this is the new return of the application token that means you can get rid of the loop over applications elsewhere, right?